### DISCUSSION TITLE
Custom attribute in EC Objects 2 (EC) and 3 (BIS)

### DISCUSSION THREAD

**Discussion by jan-slegr** [iTwin/itwinjs-core#95] [General] [answered]
I am aware of [changes in custom attributes](https://www.imodeljs.org/bis/ec/differences-between-ec2-and-ec3/) implementation in ECO2 and ECO3, but the linked page describe primarily technical aspects.

I'd like to ask whether custom attributes defined in `CoreCustomAttributes` are fixed or enhancements are planned?

When thinking about EC Objects 2 features, what I am missing:

* **`DontShowNullProperties`** is the modifier I have used often, because it allows to clean displayed data structure from occasionally filled information (e.g. description). I think there is no equivalent in BIS? Hide schema / property / class attributes are not conditional.
* **`ArrayBehaviorAttributes`** are helpful, because e.g. empty arrays in GUI looks weird, but maybe with a different approach to data modeling in BIS, a number of arrays (or even nested arrays) will be lowered and there will be no need for this attribute ;-)
* **`Category`** is replaced by PropertyCategory. Is it correct? I have not tried yet whether properties in the same category are grouped logically only or also grouped in GUI.

I am aware modeling concepts are different in former EC schemas and BIS world, so I welcome explanation how the same result can be reached in BIS (or why it's not necessary now ;-)

Regards,

  Jan

**Comment by jan-slegr** [comment]
I found another custom attribute I use quite often :-)

**`Format`** is useful e.g. when decimal number with exact decimal places should be displayed. For example it can be stored as double with a value 3.14 (always rounded to a proper number of decimal places to ensure no extra rounding will be applied when displayed) with **FormatString** set to `{0:0.00}`.

I think the answer (correct mapping from ECO2 to ECO3) is [Format class](https://www.imodeljs.org/bis/ec/ec-format/), used as [ECCustomAttribute](https://www.imodeljs.org/bis/ec/ec-custom-attributes/) at related class. Is it right?

**Comment by calebmshafer** [comment, marked as answer]
Hey @LabyrinthTechnology, the [CoreCustomAttributes](https://www.imodeljs.org/bis/domains/corecustomattributes.ecschema/) schema is not 100% fixed so if there is an idea or enhancement that is necessary/helpful and needs to be at that low of a level (rather than in BisCore or another Domain schema), it can definitely be added to future versions of the schema.

With that being said, here's the idea for the new way to handle each of the CAs you brought up: 

- `DontShowNullProperties` - The idea is to use [PresentationRules](https://www.imodeljs.org/learning/presentation/) for this type of conditional display instead of putting it directly in the ECSchema with a custom attribute.
- `ArrayBehaviorAttributes` - I believe this one is the same as above and the new place to do it is in PresentationRules
- `Category` - Yes, this has been replaced by PropertyCategory and is actually converted, while bridging, from a Category CA to a PropertyCategory.
- `Format` - You're correct the new Format class is the replacement, however those Format's are applied to a ECProperties using a [KindOfQuantity](https://www.imodeljs.org/bis/ec/kindofquantity/?term=kindof) instead of a custom attribute.

cc: @ColinKerr and @scsewall in case they have anything to add.

**Reply by jan-slegr** [reply]
Thanks a lot for the detail explanation!
Will check a description of classes you mentioned. Still a lot of things to learn :-)
